peps/pep-0705.rst

```diff
@@ -77,106 +82,308 @@ The problem disappears if we don't have mutator methods in ``Movie``. This could
+        book: Book
+
+    Entertainment = EntertainmentMovie | EntertainmentBook
+
+Users of this pattern expect type-checkers to allow the following pattern::
+
+    def get_name(entertainment: Entertainment) -> str:
+        if "movie" in entertainment:
+            return entertainment["movie"]["name"]
+        elif "book" in entertainment:
+            return entertainment["book"]["name"]
+        else:
+            # Theoretically unreachable but common defensive coding
+            raise ValueError("Unexpected entertainment type")
+
+However, type-checkers will actually raise an error on this code; mypy, for instance, will complain that ``TypedDict "EntertainmentBook" has no key "movie"`` on the third line. This is because ``TypedDict`` does not prevent instances from having keys not specified in the type, and so the check ``"movie" in entertainment`` can return True for an ``EntertainmentBook``.
+
+Users can alternatively use a non-total ``TypedDict`` instead of a union::
+
+    class Entertainment(TypedDict, total=False):
+        movie: Movie
+        book: Book
+
+This ensures the ``get_name`` example type-checks correctly, but it no longer encodes the constraint that exactly one key must be present, meaning other valid code raises spurious type-check failures. In practice, we tend to see code using types like this either casting to the correct type, with the associated risk of mistakes, or moving the ``in`` checks to dedicated ``TypeGuard`` functions, reducing readability.
+
 Rationale
 =========
 
-The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and support the cases exemplified above::
+The first two motivating examples can be solved by removing the ability to update one or more of the entries in a ``TypedDict``. This does not mean the entries are immutable; a reference to the underlying dictionary could still exist with a different but compatible type in which those entries have mutator operations. As such, these are not "final" entries; using this term would risk confusion with final attributes, which are fully immutable. These entries are "readonly".
 
-    from typing import NotRequired, TypedMapping
+To support this, we propose adding a new boolean flag to ``TypedDict``, ``readonly``, which when set to True, removes all mutator operations from the type::
 
-    class Movie(TypedMapping):
+    from typing import NotRequired, TypedDict
+
+    class Movie(TypedDict, readonly=True):
         name: str
-        year: NotRequired[int | None]
+        director: str
+
+    class Book(TypedDict, readonly=True):
+        name: str
+        author: str
+
+In addition to these benefits, by flagging arguments of a function as read-only, it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desireable property of a function interface.
```

Comment:
I first understood that this line suggests that every function that doesn't assign to its arguments should now be written like this:
```py
def foo(arg1: ReadOnly[int], arg2: ReadOnly[list[str | None]], arg3: ReadOnly[SomethingElse]) -> None: ...
```
I hope this is not the case, and it would be good to clarify that specifically using read-only TypedDict types for arguments can be useful. 

Also I'm not sure if it's explicitly spelled out that `ReadOnly` is only valid when specifying TypedDict item types, and not for attributes of normal classes or variables. It should be rejeted in type aliases as well, I think -- it's not a type modifier, but a modifier for a TypedDict item.

Reply:
I don't want to seem like using it elsewhere is ruled out in future, though. For instance, Jelle suggested it would be useful in Protocols. I can imagine it being useful on base class attributes too, as an alternative to adding a `@property` decorator. How would you suggest wording this?

Reply:
The wording looks better, thanks!

I agree that `ReadOnly` could be useful in protocols and perhaps ABCs. Maybe say something like this: "this PEP only proposes making ReadOnly valid in a TypedDict, but a possible future extension would be to support it in additional contexts, such as in protocols".

Reply:
Done, thanks :)